Add cuBLAS wrappers for tile operations

msimberg opened the pull request:
This adds the basic wrappers that do argument conversion (blaspp enums to cuBLAS enums, `std::complex` to `cuComplex`/`cuDoubleComplex`), and tests.

Questions:
- Namespaces and names? The blaspp wrappers are in `dlaf::tile`. The signatures won't be the same since the cuBLAS wrappers will take a cuBLAS handle. Is this functionality meant to be public? I've currently prefixed the wrappers with `cublas_`.
- Do you rely only on `-Wswitch-enum` to cover all cases in a switch statement? I currently added a false assertion on the default case, but I see that you haven't done that elsewhere.

Includes #289.

teonnik commented on include/dlaf/util_cublas.h:
```diff
@@ -0,0 +1,110 @@
+namespace gpu {
+
+// TODO: Are these necessary?
+inline constexpr cublasSideMode_t blas_to_cublas_side(blas::Side side) {
```
Perhaps you can overload `blas_to_cublas()` and remove the suffixes `_side`, `_uplo`, etc. 

Also an alternative name for the function could be `blas2cublas()`, in case someone prefers it. I am OK with the current name too.

msimberg replied:
I'm indifferent about the particular name. Is there precedent for `x2y`? The `_type` suffix is indeed unnecessary, thanks! I'll remove that.

msimberg replied:
I've removed the suffixes for the enumeration conversion functions. In principle `blas_to_cublas_cast` could also do without the `_cast` suffix. I slightly prefer to have it there to explicitly say that we're casting and returning the same underlying value in a different form, rather than a new value, but I can just as well go without the suffix as well.

teonnik replied:
> I've removed the suffixes for the enumeration conversion functions. In principle blas_to_cublas_cast could also do without the _cast suffix. I slightly prefer to have it there to explicitly say that we're casting and returning the same underlying value in a different form, rather than a new value, but I can just as well go without the suffix as well.

I also think it's worth keeping the suffix for that one.

> I'm indifferent about the particular name. Is there precedent for x2y

I think that's some UNIX-sm I may have subconsciously picked up (utilities like pdf2svg, etc). The current name is pretty clear, let's just keep it as it is.

teonnik reviewed (approved):
LGTM! Just a minor comment.

EDIT:

> Namespaces and names? The blaspp wrappers are in dlaf::tile. The signatures won't be the same since the cuBLAS wrappers will take a cuBLAS handle. Is this functionality meant to be public? I've currently prefixed the wrappers with cublas_.

Yes, I think the handle should be in the API in case a user has multiple handles (e.g. when using multiple GPUs from a single process).

I am not sure about the namespace, I use `dlaf::cuda::` and `dlaf::cublas::` in #234 .

> Do you rely only on -Wswitch-enum to cover all cases in a switch statement? I currently added a false assertion on the default case, but I see that you haven't done that elsewhere.

It seems we do: `cmake/DLAF_AddTargetWarnings.cmake` defines `target_add_warnings()` which includes the switch. The function is used in `miniapp/CMakeLists.txt` to include the flags to the `DLAF` target.

msimberg commented:
> LGTM! Just a minor comment.
> 
> EDIT:
> 
> > Namespaces and names? The blaspp wrappers are in dlaf::tile. The signatures won't be the same since the cuBLAS wrappers will take a cuBLAS handle. Is this functionality meant to be public? I've currently prefixed the wrappers with cublas_.
> 
> Yes, I think the handle should be in the API in case a user has multiple handles (e.g. when using multiple GPUs from a single process).

My question was confusingly worded. It was not about having the handle in the signature (that I take as a given), but rather about the `internal` or non-`internal` namespace. The blaspp wrappers are now in `dlaf::tile`, i.e. public. `dlaf::cuda::internal` is obviously not meant for public use, and the scheme doesn't fit where the blaspp wrappers are. I'd think either the blaspp wrappers should move to `dlaf::something::internal` or the cuBLAS wrappers should move to `dlaf::tile`. Preferences? I can also leave things as they are and leave it for another PR.

> I am not sure about the namespace naming, I use `dlaf::cuda::` and `dlaf::cublas::` in #234 .

Indeed, that's why I followed the same namespace.

Sely85 commented on include/dlaf/cublas_tile.h:
```diff
@@ -0,0 +1,202 @@
+};
+
+template <class T>
+void cublas_trsm(cublasHandle_t handle, blas::Side side, blas::Uplo uplo, blas::Op op, blas::Diag diag,
+                 T alpha, const matrix::Tile<const T, Device::GPU>& a,
+                 const matrix::Tile<T, Device::GPU>& b) {
```
A couple of weeks ago we added `const` to all arguments of LAPACK and BLAS wrappers (see #276). I actually don't know how cuBLAS works, sorry. But do you think it is worth marking them as `const` here too?

msimberg replied:
Yeah, absolutely, thanks!

msimberg replied:
Updated.

msimberg commented:
> > I am not sure about the namespace naming, I use `dlaf::cuda::` and `dlaf::cublas::` in #234 .
> 
> Indeed, that's why I followed the same namespace.

Uhh, I thought I did :D but I did not... I'll change this once there's a concensus on where the functionality should live.

rasolca commented:
> > LGTM! Just a minor comment.
> > EDIT:
> > > Namespaces and names? The blaspp wrappers are in dlaf::tile. The signatures won't be the same since the cuBLAS wrappers will take a cuBLAS handle. Is this functionality meant to be public? I've currently prefixed the wrappers with cublas_.
> > 
> > 
> > Yes, I think the handle should be in the API in case a user has multiple handles (e.g. when using multiple GPUs from a single process).
> 
> My question was confusingly worded. It was not about having the handle in the signature (that I take as a given), but rather about the `internal` or non-`internal` namespace. The blaspp wrappers are now in `dlaf::tile`, i.e. public. `dlaf::cuda::internal` is obviously not meant for public use, and the scheme doesn't fit where the blaspp wrappers are. I'd think either the blaspp wrappers should move to `dlaf::something::internal` or the cuBLAS wrappers should move to `dlaf::tile`. Preferences? I can also leave things as they are and leave it for another PR.
>
Tile operations should be available to users (with a public API), in case they have to develop custom algorithms.
I would use `dlaf::tile`.

rasolca commented:
Please note that function names are supposed to be lowerCamelCase. (see misc/coding.md)

msimberg commented:
I moved the cuBLAS wrappers to `dlaf::tile`. Looking around a bit more I thought `dlaf::util` might even fit better for the conversion functions. I moved them there now, but again I have no real preference (except no nesting...). `dlaf::cublas` also works.

Sely85 commented on test/unit/test_cublas_tile/test_herk.h:
```diff
@@ -0,0 +1,103 @@
+  copy(a, a0d);
+  copy(c, cd);
+
+  Tile<CT, Device::GPU> ad(std::move(a0d));
```
Do you think it would be possible, in case it is useful, to extend the usage of `createTile` to deal with GPU matrices/tiles? 
I a thinking about these functions:
https://github.com/eth-cscs/DLA-Future/blob/c8bca01662d7aad4eccc03a76272494694e72897/test/include/dlaf_test/matrix/util_tile.h#L77-L98
and 
https://github.com/eth-cscs/DLA-Future/blob/c8bca01662d7aad4eccc03a76272494694e72897/test/include/dlaf_test/matrix/util_tile_blas.h#L59-L72

msimberg replied:
I think that could be useful, yes! If you don't mind I'd like to look at that in a separate PR though. Would that be ok?

Sely85 replied:
Yes, sure!

msimberg commented:
bors try

msimberg commented:
bors try

albestro reviewed (approved):
LGTM

albestro commented on include/dlaf/cublas_tile.h:
```diff
@@ -0,0 +1,200 @@
+void cublasGemm(cublasHandle_t handle, const blas::Op op_a, const blas::Op op_b, const T alpha,
+                const matrix::Tile<const T, Device::GPU>& a, const matrix::Tile<const T, Device::GPU>& b,
+                const T beta, const matrix::Tile<T, Device::GPU>& c) {
+  SizeType m;
+  SizeType k;
+  if (op_a == blas::Op::NoTrans) {
+    m = a.size().rows();
+    k = a.size().cols();
+  }
+  else {
+    m = a.size().cols();
+    k = a.size().rows();
+  }
+  SizeType k2;
+  SizeType n;
+  if (op_b == blas::Op::NoTrans) {
+    k2 = b.size().rows();
+    n = b.size().cols();
+  }
+  else {
+    k2 = b.size().cols();
+    n = b.size().rows();
+  }
```
Just a minor thought.

I'm wondering if in the future wouldn't be worth to factor out this kind of checks, that in general are not device-dependent.

msimberg replied:
Good idea. Let me also follow up on this in a separate PR?

albestro replied:
Absolutely! 😉

rasolca commented on test/unit/test_cublas_tile/test_trsm.h:
```diff
@@ -0,0 +1,100 @@
+
+using dlaf::util::size_t::mul;
+
+template <typename T>
+void print_tile(Tile<T, Device::CPU> const& t) {
+  auto s = t.size();
+  auto n = s.cols();
+  auto m = s.rows();
+  for (int i = 0; i < m; ++i) {
+    for (int j = 0; j < n; ++j) {
+      std::cout << t(TileElementIndex{i, j}) << " ";
+    }
+    std::cout << std::endl;
+  }
+}
```
Is this debug code supposed to me added to master?

msimberg replied:
Nope, good catch!

msimberg replied:
Removed.

msimberg commented:
bors try

msimberg commented:
bors try-

msimberg commented:
bors try
